test_torch_tensor_dag.py

```diff
@@ -0,0 +1,135 @@
+# coding: utf-8
+import logging
+import os
+import sys
+import torch
+import pickle
+import io
+
+import pytest
+
+import ray
+from ray.air._internal import torch_utils
+import ray.cluster_utils
+from ray.dag import InputNode
+from ray.tests.conftest import *  # noqa
+
+from ray.experimental import TorchTensor
+
+
+logger = logging.getLogger(__name__)
+
+
+
+@ray.remote
+class TorchTensorWorker:
+    def __init__(self):
+        self.device = torch_utils.get_devices()[0]
+
+    def send(self, shape, dtype, value: int):
+        return torch.ones(shape, dtype=dtype, device=self.device) * value
+
+    def recv(self, tensor):
+        assert tensor.device == self.device
+        return (tensor[0].item(), tensor.shape, tensor.dtype)
+
+
+
+sender = TorchTensorWorker.options().remote()
+receiver = TorchTensorWorker.options().remote()
+
+# 10MiB tensor.
+shape = (50_000_000,)
+dtype = torch.float16
+
+NUM_ITERS = 10
+
+def timeit(label, fn):
+    for i in range(NUM_ITERS):
+        start = time.perf_counter()
+        fn(i)
+        end = time.perf_counter()
+        print(label, (end - start) * 1000, "ms")
+
+def exec_ray_dag(label, sender, receiver):
+    # Test torch.Tensor sent between actors.
+    with InputNode() as inp:
+        dag = sender.send.bind(shape, dtype, inp)
+        dag = TorchTensor(dag, shape, dtype)
```

Comment:
Instead of adding a wrapper DAG node, I'm thinking it may be advantageous to set typing information on the node instead to define the static `shape` and `dtype`. This would be just one case of adding more static typing information to the DAG, e.g., could imagine `size` as a generally useful attribute too.

Syntactically this could make the code look like this instead:

```
        dag = sender.send.bind(shape, dtype, inp) \
                .with_output_type(TorchTensorType(shape, type))
        dag = receiver.recv.bind(dag)
```

I haven't looked closely at how this would change the backend implementation but I think this could simplify that as well by removing special cases.

Reply:
I thought about this but the one issue with that API is that it's not very clean for tasks that return multiple values. This isn't supported yet but I suspect we will need to support it eventually.

Reply:
Although yes, we could set use that kind of dot syntax instead of a wrapper class.

Reply:
Hmm, eventually how would you support that? If the return type is arbitrarily complex then you probably need to handle it  at runtime in a custom serialization hook instead of in the DAG. Or, the type can become complex as well.

Not sure if runtime handling is sufficient to setup accelerated transfer in all cases though, or if you need to know the type information statically ahead of time.

Reply:
For multiple return values, we can use the normal Ray `num_returns` syntax.

But yeah, the type annotations will get unwieldy for nested tensors. What I'm thinking is that for cases where the user knows the shape and dtype ahead of time, they should use the static annotation and that will guarantee 0 control plane overhead from accelerated DAGs.

If they don't know it ahead of time, we can do the custom serialization hook (we just have to make sure to do it only for DAG outputs, not for other objects serialized by the user). The serialization hook would pass the type metadata through a normal channel and start the accelerated transfer separately. It'll be slower because we need to send the type metadata synchronously, but it will be much easier to program against. With the current PR, it'd also be easy for the user to add their own type hints to the task return values, by wrapping a returned torch.Tensor with the `_TorchTensorWrapper` (currently a developer API).

Reply:
Got it. I feel like maybe annotating the function is actually the best way to go long run, like `@ray.remote(return_types=[...])`, but any interim state between that and the nodes is probably fine.

I do think that using type annotations instead of a separate node wrapper class will clean up the internal code though (avoid "tensor" special cases and rewriting).